### DISCUSSION TITLE
posting realtime data from sensors to ESP32 as an AP to a connected client

### DISCUSSION THREAD

**Discussion by bmartinga** [Mair/esp32-course#108] [Course topic requests]
Hello,
I'm using the esp32 as an AP. I've been researching how to send sensor data values to a client connected to the esp32 AP in a repeating manner, i.e. once per second. I also need the connected client to be able to send text that the esp32 will use to create sd card directories, files, etc. I'm new to client/server apps and after researching for over a month, cannot seem to find the right information as to how to do this. I've been writing embedded code for micro's for 20 years but the client/server/esp idf is fairly new to me. 
Do you have an example that shows how to set the ESP32 as a server, update sensor values in real time to this server, so that a connected client can see these values change in real time, and the client have a text box that allows it to input a text string and send that text string back to the server for processing? From the best I can tell, AJAX would probably be the best interface but I'm not sure. 
This is a production product so it has to be right so to speak.
Any information you could share would be much appreciated.
Thanks

**Comment by Mair** [comment]
Hi @bmartinga 

sending data from the browser (client) to the ESP32 can be done using HTTP protocol. I cover this in the course but if you need some basic information on the http protocol I did [this video](https://www.youtube.com/watch?v=44Kqyphvlxk) that might help.

As far as updating the browser (chip to browser) is concerned and doing it in real time, you have 2 options. 
1. Use HTTP polling. 
HTTP requests always goes from the browser (client) to the server(chip) so you have to continually ask the chip for updated information on an interval. This is not ideal.

The other way is to use 

2. web sockets which is preferable. 

Web sockets allow for bi-directional communication and its very fast
As a demo of this please see [this video](https://www.youtube.com/watch?v=R9n32nxrzug) further details can be located in the course *Internet server* and the *website on chip* sections of the course.

These section assume you have been going in sequence so you might need to start at "Internet connection" and move forward if you find assumption about your knowledge being made.

Hope this helps
